### ISSUE TITLE
Performance: Write HTTP payloads once to disk, improve digest writes

### ISSUE THREAD

**Issue by csjx** [NCEAS/metacat#1380]
While troubleshooting larger file uploads (10-100+ GB range), @vchendrix noted that uploads to Metacat take much longer than uploads to the ESS-DIVE package service.  We need to track down why this is the case.
After looking at the way handle MIME/Multipart messages, we are writing the entire message to temporary disk space. We then parse the MMP file to get the`SystemMetadata` param part so we can use the `SystemMetadata.checksum` value for validating the object being written to the final resting place (usually `/var/metacat`).
Writing the object to disk twice incurs a performance hit. 
- [ ] Determine if we can write the `object` file part to `/var/metacat` once, and after it is written, compare the computed checksum to the `SystemMetadata.checksum` value after the fact.  If they match, keep the file, if not, delete it and throw an exception.
- [ ] Determine if we can increase performance of the write operation to the `DigestOutputStream` that calculates the checksum on the fly.  We do this in `D1NodeService.writeStreamToFile()` around line 1687, and in `DocumentImpl.writeToFileSystem()` around line 1563. The digest write using `IOUtils.copyLarge` looks to incur a significant performance cost. Testing results are in the following table:

File size | No Digest Duration | MD5 Digest Duration | SHA 1 Digest Duration
-------- | -------------------| --------------------- | ----------------------
  1 GB    | 0 m 02 s |  0 m 06 s    |  0 m 05 s  
 10 GB    | 0 m 28 s |  0 m 60 s    |  0 m 56 s  
100 GB    | 4 m 54 s | 10 m 36 s    | 13 m 39 s

When writing to a `DigestOutputStream` (either `MD5` or `SHA 1`) we at least double the time it takes to write the file. Is there a more efficient way to do this, and still be able to verify the correct bytes are written?

**Comment by vchendrix** [comment]
Thanks for the write-up @csjx.  I want to point out that we want to reduce the time that the client is waiting for a response from metacat so that we don't have to increase the server timeouts so much.  Minimizing this wait time ( after the client has sent  all of the bytes  to metacat ) by writing the object once and calculating the checksum as the object bytes are being written to disk as describe in your writeup is key.  Increased throughput is of course is desireable but I think it is the waiting for the response that is the focus. 

@csnavely Do you have anything to add?

**Comment by csjx** [comment]
Yes, @vchendrix - makes sense.  Writing to disk once will surely improve our response time. I'm mainly concerned that the `DigestOutputStream` is giving us a 3x performance hit for a 100 GB file, but I don't know of an alternative.  I've read some Stack Overflow discussions about how the digest is implemented under the hood and could be faster, but the solutions seems to potentially compromise integrity. I think it may need some more research.

**Comment by vchendrix** [comment]
@csjx, @csnavely and I discussed that we would be happy with an asynchronous check of the new files if it means faster ingest. This could be optional.

**Comment by mbjones** [comment]
Chris -- fully agree we should eliminate redundant writes and improve performance as much as possible.  I'm not sure if there is a more efficient approach, but if you can find one I would be all for it in the synchronous method call.

We've discussed async uploads and checksum checks in the past.  The problem is it prevents us from rejecting content for which the checksum doesn't match, which has been really important over time to avoid accidentally ingesting the wrong content or malformed content as if it is valid.  Because we require both immutability and persistence (e.g., no true *delete* function), it is really important to check that the uploaded bytes are the intended ones from the client before indicating success.  This also gives us the baseline to be sure we've got the right bytes before replication to other nodes starts.  So, to go to an async create/update API model, I think we would have to create a new API  that would involve an async checksum verification, and not make content available until the checksum validation was completed.  Otherwise, we could have clients accessing invalid content, and/or we might try to replicate invalid content.  An async API is certainly possible albeit more complicated for both the server and client, but it could have some benefits as well in being able to use out-of-band transfers over other protocols like gridftp and globus too. If that is of interest, lets discuss it in another ticket and just deal with performance on the synchronous API call here in this one.

**Comment by amoeba** [comment]
Nice writeup, @chris. A few thoughts:

1. From your OP, it sounds like part of this slowness may be due to needing to parse the MMP message first in order to grab the system metadata. If we can't work around that, I think it'd be reasonable to allow clients to specify a checksum (and algo) in the HTTP header of the POST request, which the receiving client would use preferentially to verify the upload
2. As far as algorithms to verify content go, MD5 is slow, can faster algorithms be used here?
3. I didn't have a prior on how fast an HTTP server should be able to receive and hash a file so I put a simple Node.js (Express) server together that writes to disk only once and calculates the checksum as it writes. A 10GB file takes about 30 seconds. Just FYI.

**Comment by csnavely** [comment]
Chiming in late due to conference / vacation. The big problem is that the checksum creation is a) not in-band with the transfer (ie a separate post-process) and b) linear with the file size.

Put another way, slow transfer is workable because it's not inactivity. With the current design, we have to set inactivity timeouts in every tier that the traffic passes through, essentially matching the long period of silence during the checksum creation in post-processing.

**Comment by mbjones** [comment]
Discussed this today on the devteam call.  Seems like we need to parse the stream and calculate the checksum as it is transferred over the wire.  Here's a streaming multipart parser: http://commons.apache.org/proper/commons-fileupload/streaming.html
Because that provides an InputStream for each body part, it seems like we could pipe it straight into the checksum routine so the mime part is read and checksummed while it is being mime decoded. Which I think might mean that it would be processed during the network transfer, and so may keep the connection open as it goes.

**Comment by taojing2002** [comment]
Neither Chris nor I could find a better way to calculate the checksum than the DigestOutputStream class.

**Comment by mbjones** [comment]
@taojing2002 Is this streaming checksum approach working now? I just saw this ticket, and realized we were talking about it. If it isn't, then we should just fold this ticket into the HashStore refactor under epic #1597.

**Comment by taojing2002** [comment]
@mbjones It was done and it has worked for while. Somehow this ticket wasn't closed.
